src/plugins/console/public/lib/autocomplete/autocomplete.ts

```diff
@@ -1123,80 +1126,105 @@ export default function ({
+      }
+      if (t1.name === t2.name) {
+        return 0;
+      }
+      return 1;
+    });
+
+    return terms;
+  }
+
+  function getSuggestions(terms: ResultTerm[]) {
+    return _.map(terms, function (t, i) {
+      t.insertValue = t.insertValue || t.value;
+      t.value = '' + t.value; // normalize to strings
+      t.score = -i;
+      return t;
+    });
+  }
+
   function getCompletions(
     position: Position,
     prefix: string,
     callback: (e: Error | null, result: ResultTerm[] | null) => void
   ) {
     try {
       const context = getAutoCompleteContext(editor, position);
+
       if (!context) {
         callback(null, []);
       } else {
-        const terms = _.map(
-          context.autoCompleteSet!.filter((term) => Boolean(term) && term.name != null),
-          function (term) {
-            if (typeof term !== 'object') {
-              term = {
-                name: term,
-              };
-            } else {
-              term = _.clone(term);
-            }
-            const defaults: {
-              value?: string;
-              meta: string;
-              score: number;
-              context: AutoCompleteContext;
-              completer?: { insertMatch: (v: unknown) => void };
-            } = {
-              value: term.name,
-              meta: 'API',
-              score: 0,
-              context,
-            };
-            // we only need our custom insertMatch behavior for the body
-            if (context.autoCompleteType === 'body') {
-              defaults.completer = {
-                insertMatch() {
-                  return applyTerm(term);
-                },
-              };
-            }
-            return _.defaults(term, defaults);
-          }
-        );
-
-        terms.sort(function (
-          t1: { score: number; name?: string },
-          t2: { score: number; name?: string }
-        ) {
-          /* score sorts from high to low */
-          if (t1.score > t2.score) {
-            return -1;
-          }
-          if (t1.score < t2.score) {
-            return 1;
-          }
-          /* names sort from low to high */
-          if (t1.name! < t2.name!) {
-            return -1;
-          }
-          if (t1.name === t2.name) {
-            return 0;
-          }
-          return 1;
-        });
-
-        callback(
-          null,
-          _.map(terms, function (t, i) {
-            t.insertValue = t.insertValue || t.value;
-            t.value = '' + t.value; // normalize to strings
-            t.score = -i;
-            return t;
-          })
-        );
+        const terms = getTerms(context, context.autoCompleteSet!);
+
+        const suggestions = getSuggestions(terms);
+
+        if (!context.asyncResultsState?.isLoading) {
+          callback(null, suggestions);
+        }
+
+        if (context.asyncResultsState) {
+          // TODO indicate mappings fetching is in progress
+          context.asyncResultsState.results$.subscribe((r) => {
```

Comment:
Is it possible that we have a memory leak here: calling `.subscribe` on new observables without unsubscribing?

Reply:
replaced observable with a promise in https://github.com/elastic/kibana/pull/147770/commits/ea68a57143f218a5b54e61c066a31591125e7500